network/p2p/unicast/manager.go

```diff
@@ -138,49 +244,108 @@ func (m *Manager) rawStreamWithProtocol(ctx context.Context,
 		// cancel the dial back off (if any), since we want to connect immediately
 		dialAddr = m.streamFactory.DialAddress(peerID)
 		m.streamFactory.ClearBackoff(peerID)
-
-		// if this is a retry attempt, wait for some time before retrying
-		if retries > 0 {
-			// choose a random interval between 0 to 5
-			// (to ensure that this node and the target node don't attempt to reconnect at the same time)
-			r := rand.Intn(MaxConnectAttemptSleepDuration)
-			time.Sleep(time.Duration(r) * time.Millisecond)
-		}
-
 		err := m.streamFactory.Connect(ctx, peer.AddrInfo{ID: peerID})
 		if err != nil {
-
 			// if the connection was rejected due to invalid node id, skip the re-attempt
 			if strings.Contains(err.Error(), "failed to negotiate security protocol") {
```

Comment:
Inspecting the error string is at odds with our error handling conventions. I understand that we are dealing with LibP2P errors here, and they don't have proper sentinels. But this shortcoming from LibP2P is nothing we should be exposing to our business logic. 

We have the `StreamFactory` interface as a wrapper around libP2P that should encapsulate this shortcoming of LibP2P. Inspecting errors and converting them to proper sentinels could easily be done in the `StreamFactory` implementation: https://github.com/onflow/flow-go/blob/77190f4772df8d06f24e3e29b6133208ad111c60/network/p2p/unicast/streamfactory.go#L46-L48

Suggestion:
* move this into error string inspection to `StreamFactory` 
* add a sentinel to `StreamFactory` that represents this case of "invalid node id"
* document `StreamFactory.Connect` method in the interface including sentinel https://github.com/onflow/flow-go/blob/77190f4772df8d06f24e3e29b6133208ad111c60/network/p2p/unicast/streamfactory.go#L20
* Ideally, we would have a test verifying that our error conversion works (in case libP2P changes the error string at some point)
* `rawStreamWithProtocol` should also document the possibility of this error return

With these changes, `StreamFactory`  has the `Connect` method, which follows our error documentation guidelines.

Reply:
https://github.com/onflow/flow-go/pull/3902/commits/9452b35c05dbfc7b2b0c50aeb0b652200e845dae